lib/rpush/daemon/gcm/delivery.rb

```diff
@@ -72,10 +72,17 @@ def process_response(response)
         end
 
         def handle_successes(successes)
-          successes.each do |result|
-            reflect(:gcm_delivered_to_recipient, @notification, result[:registration_id])
-            next unless result.key?(:canonical_id)
-            reflect(:gcm_canonical_id, result[:registration_id], result[:canonical_id])
+          return if successes.blank?
+
+          if Rpush.config.gcm_batch_reflections
+            successes_map = successes.map { |result| [result[:registration_id], result[:canonical_id]] }.to_h
```

Comment:
It looks like potential memory issue here and for `failures_map` in `handle_errors` as well. 

`successes` can hold thousands of objects and `map` will create a new array in memory, the same `to_h` does. 

You can use something like `inject` or `each_with_object ` in order to reduce the number of iterations and objects but still not sure if it is good idea "double" the memory consumption.